src/display.jl

```diff
@@ -1,34 +1,92 @@
-using .LineNumbers: SourceFile, compute_line
-using CSTParser
-using CSTParser: Error
+# Find first parser error. All other errors are ignored.
+function first_error(ex::EXPR, pos=0)
+    _first_error(ex, Tuple{EXPR,Int}[], pos)
+end
 
-function display_diagnostic(io::IO, code, diagnostics::Vector{Error}; filename = "none")
-    file = SourceFile(code)
-    for (i,message) in enumerate(diagnostics)
-        if isempty(message.loc)
-            i != 1 && (printstyled(io, "ERROR", color=:red, bold=true); print(io, ": "))
-            println(io, message.description)
-            continue
-        end
-        offset = first(message.loc)
-        line = compute_line(file, offset)
-        str  = String(file[line])
-        lineoffset = offset - file.offsets[line]
-        col  = (lineoffset == 0 || isempty(str)) ? 1 :
-               lineoffset > sizeof(str) ? textwidth(str) + 1 :
-                textwidth(str[1:lineoffset])
-        if false #message.severity == :fixit
-            print(io, " "^(col-1))
-            printstyled(io, message.text, color=:green)
-            println(io)
-        else
-            print(io, "$filename:$line:$col " )
-            i != 1 && (printstyled(io, "ERROR", color=:red, bold=true); print(io, ": "))
-            println(io, message.description)
-            println(io, rstrip(str))
-            print(io, " "^(col-1))
-            printstyled(io, string('^',"~"^(max(0,length(message.loc)-1))); color=:green, bold = true)
-            println(io)
+function _first_error(ex, trace, pos)
+    push!(trace, (ex,pos))
+    if typof(ex) === CSTParser.ErrorToken
+        return trace
+    end
+    for i in 1:length(ex)
+        res = _first_error(ex[i], trace, pos)
+        if !isnothing(res)
+            return res
         end
+        pos += ex[i].fullspan
+    end
+    pop!(trace)
+    return nothing
+end
+
+function make_underline(width,
+                        #chars=('^','~','^','|')
+                        chars=('▔','▔','▔','▎')
+                        #chars=('┗','━','┛','▎')
+                       )
+    if width == 0
+        chars[4]
+    elseif width <= 2
+        chars[2]^width
+    else
+        string(chars[1], chars[2]^(width-2), chars[3])
+    end
+end
```

Comment:
You are probably aware but I'll mention there is `printstyled(io, str, color=:underline)`

Reply:
Yes that's an interesting option - perhaps worth doing, especially for multi-line errors. We could use underline, red,bold or some other styling. (It's even an option to have syntax highlighting combined with error reporting - Tokenize + CSTParser gives us more than enough tooling for that.)

The thing which I find nice about the box chars is that they work in black and white, and they can be colored separately from the source code.  But the need to format multi-line errors may override that consideration.

Reply:
As you can see in the screencast, I've tried in-situ highlighting in the latest version which is an easier option when the error spans multiple lines.